src/global_logic.js

```diff
@@ -1692,8 +1694,31 @@ module.exports.addSkilldataToTotals = function (totals, comb, arml, buff) {
                             totals[key]["cosmosDebuffResistance"] = comb[i] * 20.0;
                         }
-                    } else if (stype == 'cosmosArm') {
-                        // Skip Cosmos Weapons Skill
+                    } else if (stype == 'cosmosLimit') {
+                        // Cosmos Weapons Damage Limit Skill
+                        if (amount == 'sword') {
+                            totals[key]["normalDamageLimit"] += comb[i] * 0.01 * Math.min(10, epic.countSwordType(arml, comb));
+                            totals[key]["ougiDamageLimit"] += comb[i] * 0.01 * Math.min(10, epic.countSwordType(arml, comb));
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countSwordType(arml, comb)));
+                        } else if (amount == 'dagger') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countDaggerType(arml, comb)));
+                        } else if (amount == 'spear') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countSpearType(arml, comb)));
+                        } else if (amount == 'axe') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countAxeType(arml, comb)));
+                        } else if (amount == 'wand') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countWandType(arml, comb)));
+                        } else if (amount == 'gun') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countGunType(arml, comb)));
+                        } else if (amount == 'fist') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countFistType(arml, comb)));
+                        } else if (amount == 'bow') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countBowType(arml, comb)));
+                        } else if (amount == 'music') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countMusicType(arml, comb)));
+                        } else if (amount == 'katana') {
+                            totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, epic.countKatanaType(arml, comb)));
+                        }
```

Comment:
I suggest generic solution
- [ ] export genCountWeaponFunc

```js
const countWeaponType = epic.genCountWeaponFunc((arm) => arm.armType === amount);
totals[key]["cosmosNormalDamageLimit"] = Math.max(totals[key]["cosmosNormalDamageLimit"], 0.01 * Math.min(10, countWeaponType(arml, comb)));
```

This will reduce test and maintenance cost 10x
testers had to test those all 10 types weapon to ensure those code does not have mistake on copy and paste.
epic.js has unit test, unit test can guarantee if any type had pass the test then other types also ok.

- [ ] no need other countXXXTypes functions
if we wants add those functions, then need to add unit test to epic.test.js

Reply:
I rather have 10 functions from the same place near each other used in 100 different places as 100 times the same code with magic strings,,,

Reply:
that make extra maintenance and test cost to have a lot if-else branch and copy-pasted code
when we found bug then maintainer have to edit 10x places and tester check those all cases.

Reply:
about Math.max check, I don't know how it should be
are there verification source if two or more Cosmos weapon was set

current code behavior is:
- two cosmos sword -> no duplicate the effects
- cosmos sword and cosmos katana -> one of max count is used

Reply:
https://twitter.com/Hecate_mk2/status/1207917620080459776

Reply:
the behavior is ok
about the many copy-paste, you can think why it shouldn't 
it requires tests tasks each weapon types to check those behaviors.